File: typescript-javascript/day14-capstone/solution/src/storage.ts
import * as fs from "fs/promises";
import { Expense, ExpenseData } from "./models";
import { ExpenseTracker } from "./tracker";

const DEFAULT_PATH = "expenses.json";

export async function loadTracker(path: string = DEFAULT_PATH): Promise<ExpenseTracker> {
    const tracker = new ExpenseTracker();
    try {
        const contents = await fs.readFile(path, "utf-8");
        const data: ExpenseData[] = JSON.parse(contents);
        tracker.expenses = data.map((item) => Expense.fromJSON(item));
    } catch (error) {
        // file doesn't exist yet on first run -- start with an empty tracker
    }
    return tracker;
}

export async function saveTracker(tracker: ExpenseTracker, path: string = DEFAULT_PATH): Promise<void> {
    const data = tracker.expenses.map((expense) => expense.toJSON());
    await fs.writeFile(path, JSON.stringify(data, null, 2));
}

export async function exportCSV(tracker: ExpenseTracker, path: string): Promise<void> {
    const header = "amount,description,category,timestamp";
    const rows = tracker.expenses.map((expense) => {
        const data = expense.toJSON();
        return `${data.amount},"${data.description}",${data.category},${data.timestamp}`;
    });
    await fs.writeFile(path, [header, ...rows].join("\n"));
}

File: python/day14-capstone/solution/src/expense_tracker/storage.py
from __future__ import annotations

import json
import csv
from pathlib import Path

from .models import Expense
from .tracker import ExpenseTracker

DEFAULT_PATH = Path("expenses.json")


def load_tracker(path: Path = DEFAULT_PATH) -> ExpenseTracker:
    tracker = ExpenseTracker()
    try:
        with open(path, encoding="utf-8") as f:
            data = json.load(f)
        tracker.expenses = [Expense.from_dict(item) for item in data]
    except FileNotFoundError:
        pass
    return tracker


def save_tracker(tracker: ExpenseTracker, path: Path = DEFAULT_PATH) -> None:
    with open(path, "w", encoding="utf-8") as f:
        json.dump([e.to_dict() for e in tracker.expenses], f, indent=2)


def export_csv(tracker: ExpenseTracker, path: str) -> None:
    with open(path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=["amount", "description", "category", "timestamp"])
        writer.writeheader()
        for expense in tracker.expenses:
            writer.writerow(expense.to_dict())
